linera-views: Add module to support splitting large values across multiple keys (#862)
linera-io/linera-protocol: linera-views/README.md

<!-- cargo-rdme end -->

The package provides essentially two functionalities:
* An abstraction to access databases.
* Several containers named views for storing data modeled on classical ones.

See `DESIGN.md` for more details.

> ## The supported databases.

Reviewer: Remove the dots at the end of titles?

Can we move these sections to the inline doc in `lib.rs`? They will be brought back above by cargo-rdme

Author: Ok, done.

See the [CONTRIBUTING](../CONTRIBUTING.md) file for how to help out.

## Design

> # Possible clients

Reviewer: maybe `# Interfacing with key-value stores` or `# Database clients` or simply `# Key-value store clients`?

> We designed a `KeyValueStoreClient` trait that implements the basic functionalities

Reviewer: "implements" is a weird for a trait. Maybe "represents"?

Reviewer: "We have designed" is more consistent with the rest of the text in present tense (I think)

Author: Ok.

> of a ke-value store whose keys are `Vec<u8>` and whose values are `Vec<u8>`.

Reviewer: typo: key-value

Author: Yes sorry the Grammarly corrections were not upstreamed by accident.

> of a ke-value store whose keys are `Vec<u8>` and whose values are `Vec<u8>`.

Author: Yes.

> A number of key-value stores are supported by the system:

Author: We rewrite as `We provide an implementation of the trait `KeyValueStoreClient` for the following key-value stores:`.

* `memory` is using the memory (and uses internally a simple B-Tree map).

> * `RocksDB` is a disk based key-value store

Reviewer: disk-based

Author: Ok.

> * `DynamoDb` is the AWS based DynamoDb service.

Reviewer: AWS-based

Author: ok.

> * `DynamoDb` is the AWS based DynamoDb service.

Author: Ok done. Also renamed in the code.

> For each of those we provide a type that implements the trait.

Reviewer: Maybe move this sentence up and replace the one before the enumeration:
`We provide an implementation of the trait `KeyValueStoreClient` for the following key-value stores:`

Author: ok.

> There are other possible clients that could be added: Redis, Scylladb, Cassandra, etc.

Reviewer: Words like "possible", "could", "may", "will", etc convey a sense of uncertainty, but in technical documentation, we're mostly interested in facts.

Here, I would not suggest a particular roadmap to add Redis, etc. We can just remove the sentence or express actual design goals:
"The trait KVSC was designed so that more storage solutions can be easily added in the future."

Author: Ok, we have several plans. But yes, let us write that only in the GitHub Issues.

> There are other possible clients that could be added: Redis, Scylladb, Cassandra, etc.

Author: We rewrote as:
"The trait `KeyValueStoreClient` was designed so that more storage solutions can be easily added in the future."

The `KeyValueStoreClient` trait is also implemented for several internal constructions of
clients:
* The `LruCachingKeyValueClient` client that implements the Least Recently Used (LRU)
caching of reads into the client.
* The `ViewContainer<C>` client that implements a key-value store client from a context.
* The `KeyValueStoreClientBigValue<K>` client that implements a client for which the
size of the values is unbounded (some key-value store clients have strict limitations

> on the value size)

Author: Ok.

# Views

A view is a container that mimics some of the properties of classic containers such

> as `VecDequeue`, `BTreeMap`, etc. with the following goals:

Author: Yes, you are right. Though maybe rust should have used `VecDequeue`, `Vector` instead of `Vec` and so on.

* The data structure can be written to the key-value store and loaded from there.
* The changes are first done in memory and accessed there. The changes can be committed
to the key-value store client when adequate.

There are various views corresponding to a number of use cases. All of those are

> implementing by mapping the values in their specific types to their serializations.

Author: Yes.

# Allowed keys, base_key and prefixes

The keys are constructed step by step from the prefix to the full key. So, whenever
we have a view or any other object then we have a `base_key` associated to it and

> to no other types.

Author: Thanks for those questions, I replied to your question in the text.

> This lead us to introduce a trait `Context` that can be implemented with a specific

Author: Ok for leads.

`KeyValueStoreClient` client and a `base_key`.

Another issue to consider is that we do not want to just store the values, but we
also need to accomodate other features:
* Administrative data (like journal) need to store some data starting from the `base_key`.
* Some of the values need to be split into several smaller values.

The rule for constructing are the following:
* For the construction of `struct` objects of associated base key `bk` we do the
following: If the corresponding type has entries `entry_0`, ..., `entry_k` then the

> base key of the object associated to the `k`-th entry is `[bk * * * *]` with `[* * * *]`

Author: Ok.

the serialization of the number `k` considered as a `u32`.
* For each view type, we have some associated data. Some are the hash, others the counts
(e.g. for `LogView` and `QueueView`) and some the index. The corresponding enum is
named `KeyTag` for each of those containers. For the index, the values are serialized.
* For a given base key `base_key` the keys of the form `[base_key tag]` with tag in
`0..MIN_VIEW_TAG` are reserved for administrative data.

These construction are done in order to satisfay the following property: The set of keys
of a view is prefix-free.

## Chosen implementation

> # Implementation of the `keyValueStoreClient`.

Author: Yes, API is better.

The design of the key value store client is done in the following way:
* We can put a `(key, value)` and delete a `key`.
* We can retrieve the list of keys having a specific prefix. Similarly, we can

> retrieve the list of `(key, value)` whose key have a specific prefix.

Author: Done.

* Also, the list of the keys are provided in the lexicographic order.

# The journaling.

> We want to make sure that all operations can be done on the database in a sure

Author: Yes, that is the correct terminology.

> way. That is the operation have to be processed completely. The idea is to

Author: Ok.

> store the journal in the `[base_key 0 * * * *]` keys with `[* * * *]` the

Author: Ok, I added documentation on that aspect.

serialization of an `u32` entry. Once written, the journal is processed
until all entries have been treated.

> Those keys are not overlaping with the other keys since their value is of the

Author: You are right, colliding is better.

form `[base_key tag *]` with `tag >= MIN_VIEW_TAG` and currently we have
`MIN_VIEW_TAG = 1`.

## The `KeyValueStoreClientBigValue` type.

> Each key value store client has a maximal size of its values (named `MAX_VALUE_SIZE`
> in the code). If we want to have unlimited size, then we can build a
> key-value store client with unlimited size.

Reviewer 2: ```suggestion
Some key-value store clients limit the size of the values (named `MAX_VALUE_SIZE`
in the code). `KeyValueStoreClientBigValue` is a wrapper that accepts values
of any size. Internally, it splits them into smaller pieces and stores them
using a wrapped, possibly size-limited, client.
```

Author: Ok.

The built key-value store client has `MAX_VALUE_SIZE` set to its maximum possible
value. This forces us to split the value into several smaller values. One key
such example is `DynamoDb` which has a limit of 400kB on its value size.

Design is the following:
* For every `key` we build several corresponding keys `[key * * * *]` that contain

> each a piece of the full data set.

Reviewer 2: ```suggestion
each a piece of the full value.
```

Author: Yes.

The `[* * * *]` is the serialization of the index which is `u32` so that
the order of the serialization is the same as the order of the `u32` entries.

> * For the first key, that is `[key 0 0 0 0]` the information is `[* * * * value_0]`

Reviewer 2: It's confusing to use terms like "data set", "information" and "value" interchangeably. Maybe we should establish a terminology instead to distinguish between the large, logical, high-level value, and the small, low-level, stored values, before explaining how one the former gets split up into the latter.

Author: Ok, the "information" and "data set" have been removed.
Regarding the large, logical, high-value, I am not completely sure what you mean and maybe we do not need such terminology.

Reviewer 2: I mean the large value, that we can't store.
As opposed to the small values that we do store and that represent the large value's chunks.

and `value_0` the first block of the value. For the other keys `[key * * * *]` the
corresponding value is `value_k` with `value_k` the k-th entry in the value. The full
value is reconstructed as `value = [value_0 value_1 .... value_N]`
The number of blocks is always non-zero, even if the value is empty.

The effective working of the system relies on several design choices.
* In `find_keys_by_prefix` and similar, the keys are
returned in lexicographic order.
* The count and indices are serialized as `u32` and the serialization is done
so that the `u32` ordering is the same as the lexicographic ordering. So, we
get first the first key (and so the count) and then the segment keys one by
one in the correct order.
* Suppose that we have written a `(key, value_1)` with several segments and then
we replace by a `(key, value_2)` which has fewer segments.
Our choice is to leave the old segments in the database. Same with delete:
we just delete the first key.
* We avoid the overflowing of the system by the following trick: When we delete
a view, we do a delete by prefix, which thus deletes the old segments if present.

> * In order to have a segment that deletes all the unused segments, every write needs

Reviewer 2: ```suggestion
* In order to delete all the unused segments when overwriting a value, every write would need
```

Author: Ok.

to have a read just before which is expensive.

> key difficulty of this modelization is that while deleting a key of the form
> `[key 0 0]` we would also delete segments of the key of the form `key`.

Reviewer 2: ```suggestion
A difficulty with this approach is that while deleting a key of the form
`[prefix 0 0]` we would also delete segments of a value stored under the key
`prefix`.
```

…but maybe that's too much information for a README? Otherwise maybe it needs an even more detailed explanation.

See the [CONTRIBUTING](../CONTRIBUTING.md) file for how to help out.

## Design

> # Key Value Store Clients

Reviewer: `###`

Author: I choose to use `#` in coherence with the `README.md`.

The keys are constructed step by step from the prefix to the full key. So, whenever
we have a view or any other object then we have a `base_key` associated with it and

> to no other objects. A view is associated to many keys, all of whom share the same

Author: Done.
